### ISSUE TITLE
writeVcf save invalid vcf with metadata out of order

### ISSUE THREAD

**Issue by apastore** [Bioconductor/VariantAnnotation#15]
If I read a valid vcf with the last 3.8 Bioconductor version 1.28.1
```
param_T <- ScanVcfParam(sample = samples_names[1],  geno = c("AD", "MQ", "DP", "AF"), info = c("DP","AF", "CALLERS"))
  Tumor <- readVcf(file = vcf_file[i], genome = "GRCH37", param = param_T)
  

```

```
head of  vcf_file[i
##fileformat=VCFv4.2
##contig=<ID=1,length=249250621>
##contig=<ID=2,length=243199373>
##contig=<ID=3,length=198022430>
##contig=<ID=4,length=191154276>
##contig=<ID=5,length=180915260>
##contig=<ID=6,length=171115067>

```

and then I save it with 

`writeVcf(Tumor, paste0("./vcf/",samples_names[1], ".vcf"), index = F)
`  

the header is out of order ##fileformat=VCFv4.2 is not on the first line and there is a Strange SAMPLE field 

Thanks for you support!

```
##SAMPLE=Tumor
##commandline="/home/pastore/data/bcbio/anaconda/bin/freebayes -f /lila/data/abdelwao-lab/pastorea/bcbio/genomes/Hsapiens/GRCh37/seq/GRCh37.fa --genotype-qualities --strict-vcf --ploidy 2 --targets /lila/data/abdelwao-lab/pastorea/Project/Ben/histiocytosis/hempact/work/freebayes/1/B02_816_T1-1_0_249250621-regions-nolcr.bed --min-repeat-entropy 1 --no-partial-observations --min-alternate-fraction 0.05 --pooled-discrete --pooled-continuous --report-genotype-likelihood-max --allele-balance-priors-off /lila/data/abdelwao-lab/pastorea/Project/Ben/histiocytosis/hempact/work/align/C_K0Y3K8_T001_d/C_K0Y3K8_T001_d-sort.bam /lila/data/abdelwao-lab/pastorea/Project/Ben/histiocytosis/hempact/work/align/PON/PON-sort.bam"
##fileDate=20181115
##fileformat=VCFv4.2
##phasing=none
```

**Comment by FreekManders** [comment]
I have the same issue. The order of the headerlines is wrong and the ALT line doesn't contain a id. It becomes:
```
##ALT=Represents any possible alternative allele at this location
```
instead of

```
##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">
```
I think the problems might be with the makeVcfHeader function, which seems to contain some broken code.

**Comment by vobencha** [comment]
@apastore @FreekManders I believe this was fixed with these two commits.
```
commit caaaf75eb2004e04ad92eeadfb568d8992a8ad2a 
Author: vobencha <valerie.obenchain@roswellpark.org>
Date:   Tue Nov 27 12:21:54 2018 -0800

    Fix bug in writeVcf() when 'fileformat' is missing

commit 21c508e83c013b1655789cb832217609d2602e4b
Author: vobencha <valerie.obenchain@roswellpark.org>
Date:   Mon Nov 19 12:33:03 2018 -0800

    writeVcf() now writes 'fileFormat' as the first header line
```
Please reopen the issue if you're still having problems.

**Comment by vobencha** [comment]
Hi @FreekManders 

Installing packages manually from github is generally not the best approach. We have both an active release (R 3.5.1; Bioconductor 3.8) and devel branch (R 3.6.*; Bioconductor 3.9). If you install manually from github you must be sure you're pulling for the appropriate branch for your version of R. Additionally, the github version often lags behind the version on the canonical git server, git.bioconductor.org.

From your sessionInfo() (thanks for posting) you now have the devel version of VariantAnnotation in your release version of R.

The best way to install Bioconductor packages is to instead use the BiocManager package (replaced the BiocInstaller package). The package will detect your version of R and install the appropriate Bioconductor package versions.
```
BiocManager::install("VariantAnnotation")
```
The release version of VariantAnnotation you should be using is 1.28.3 as seen on the release build report:

https://www.bioconductor.org/checkResults/release/bioc-LATEST/

I've tested with VariantAnnotation 1.29.9 (devel) and 1.28.3 (release) using the structural.vcf from the package. In both cases I see 'fileformat' at the top of the output file and ALT is properly written with the ID field. 
```
> fl = system.file("extdata", "structural.vcf", package="VariantAnnotation")
> vcf = readVcf(fl)
> tmp = tempfile()
> writeVcf(vcf, tmp)
> tmp
[1] "/tmp/RtmpdHJDVS/file26b16202f358"
```
The vcf specs do not specify order in the header fields except to say that fileformat must be first.
https://samtools.github.io/hts-specs/VCFv4.2.pdf

The output from the example above does list 'fileformat' first. All major categories are grouped together but header lines such as 'assembly' or 'reference' may not be in slightly different order as the original file. To my knowledge this is not a problem. 

If you still see a problem with the ALT field, please provide a small sample file so I can reproduce the issue.

Thanks.

**Comment by FreekManders** [comment]
Hi @vobencha 

Thank you for your helping me. I had switched to the github version of the package, because I thought that version might be newer and fix the issue I was having. I have now removed the package and re-installed it with the BiocManager package. 

However, I'm still having the same issue. I attached a modified version of the structural.vcf from the package, which contains the ALT field, as it is present in my vcfs. I also attached the file I get when reading and writing out this example vcf with *VariantAnnotation*. (I changed the extensions to .txt so I could attach the files)
[test.txt](https://github.com/Bioconductor/VariantAnnotation/files/2732443/test.txt)
[test_out.txt](https://github.com/Bioconductor/VariantAnnotation/files/2732444/test_out.txt)

As you can see, the ALT field is altered in the output. This alteration results in an error, when trying to read in the file in IGV:
````Unable to parse header with error: Invalid VCFSimpleHeaderLine: key=ALT name=null````

p.s. I noticed that BiocManager installed version 1.28.5 of the package, even though I'm still on R version 3.5.1. However, I don't think this is causing my issue, because I've had the same issue with other versions of the package.

**Comment by vobencha** [comment]
@FreekManders Thanks for the reproducible example and good catch! This is now fixed in 1.28.7 and 1.29.12. These versions should be available via BiocManager::install() ~noon EST tomorrow.
Valerie
